### ISSUE TITLE
[QST] What is the expected memory overhead of creating dmatrix using cudf.DataFrame

### ISSUE THREAD

**Issue by VibhuJawa** [dmlc/xgboost#4917]
#### [QST] What is the expected memory overhead of creating dmatrix using cudf.DataFrame

**Q1.** 
In my experiments i am seeing a 1.7x-2x memory overhead of `DMatrix` creation i.e `DMatrix` is approximately using double the memory as a `cudf.DataFrame`. Is this expected behavior ? 

**Q2.** 
If this is indeed expected behavior and the `dmatrix` objects indeed to require 1.7-2x the total memory,  Is there a possible way to release memory from the `cudf.DataFrame` to provide some memory relief ?  

A naive idea (probably un-usable) idea that comes to mind, is some how iterating via cudf columns,  creating a `dmatrix` column for that particular,  dropping the column in `cudf.DataFrame`  and then concatenating all `column-dmatrix ` objects. 

#### Memory overhead Example:

```python
import cudf
import xgboost
import numpy as np

df = cudf.DataFrame({str(i):np.ones(20_000_000,dtype=np.int32) for i in range(0,40)})
```

#### Memory after `dataframe` creation:

```
!nvidia-smi
Mon Oct  7 16:35:02 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.48                 Driver Version: 410.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-SXM2...  On   | 00000000:06:00.0 Off |                    0 |
| N/A   35C    P0    48W / 300W |   3477MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      5902      C   .../vjawa/anaconda3/envs/RAPIDS/bin/python  3467MiB |
+-----------------------------------------------------------------------------+
```

#### Memory after creating `DMatrix` from df

```python
dmat = xgboost.DMatrix(df)
!nvidia-smi
```

```
Mon Oct  7 16:35:05 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.48                 Driver Version: 410.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-SXM2...  On   | 00000000:06:00.0 Off |                    0 |
| N/A   35C    P0    48W / 300W |   9745MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      5902      C   .../vjawa/anaconda3/envs/RAPIDS/bin/python  9735MiB |
+-----------------------------------------------------------------------------+
```

#### Memory after deleting df
```python
del df
!nvidia-smi
```
```
Mon Oct  7 16:35:12 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.48                 Driver Version: 410.48                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-SXM2...  On   | 00000000:06:00.0 Off |                    0 |
| N/A   35C    P0    48W / 300W |   6625MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

```

**Comment by trivialfis** [comment]
> In my experiments i am seeing a 1.7x-2x memory overhead of DMatrix creation i.e DMatrix is approximately using double the memory as a cudf.DataFrame. Is this expected behavior ?

Yes.  XGBoost copies the data into a CSR format.

>  Is there a possible way to release memory from the cudf.DataFrame to provide some memory relief ?

A non-hacky way, maybe in the future when we can specialize DMatrix for histogram based algorithm.

> ropping the column in cudf.DataFrame and then concatenating all column-dmatrix objects

@wbo4958 Here beat you to it. ;-)  But right now I don't think we can merge this into XGBoost.

**Comment by wbo4958** [comment]
Hi VibhuJawa,

**For Q1**
Yeah, the issue is expected. That's because DMatrix will hold index and fvalue, So basically double CUDF size if the data is dense.

**For Q2**
Yes, the same idea as we did, pls refer https://github.com/rapidsai/xgboost/tree/rapids-spark.
But we did this in scala layer, you know, feeding chunk size to CUDF reading, and then concatenated
to DMatrix.

I strongly recommend supporting the Dmatrix building incrementally, for we can fully leverage GPU memory. @trivialfis
